pkg/monitortests/testframework/uploadtolokiserializer/monitortest.go

```diff
@@ -163,50 +191,134 @@ func UploadIntervalsToLoki(intervals monitorapi.Intervals) error {
+		if err != nil {
+			return err
+		}
 	}
 
 	logrus.Info("bearer token obtained")
 	client := &http.Client{}
 
-	// NOTE: there is a promtail go client available at https://github.com/grafana/loki/blob/main/clients/pkg/promtail/client/client.go,
-	// however it appears near impossible to vendor. For now, we roll our own with http requests, the worst
-	// part of which is we lose out on resilient retries in the event the ingesters are overloaded.
-
-	values := make([][]interface{}, 0)
 	// This format matches the labels used for the cluster logs in loki, and we want this
 	// to be labelled the same.
 	invoker := fmt.Sprintf("openshift-internal-ci/%s/%s", jobName, buildID)
 	if invoker == "" {
 		return fmt.Errorf("OPENSHIFT_INSTALL_INVOKER is needed for the invoker loki label")
 	}
-	labels := map[string]string{
-		"invoker": invoker,
-		"type":    "origin-interval",
-		// TODO: in future, may group intervals in origin rather than in the js itself, if we do,
-		// ensure this group gets carried over to this series.
-	}
+	// maps namespace name to all the log entries for that namespace. used as we need to batch up all the log entries
+	// per stream (set of labels), which includes namespace. Any intervals without a namespace, or in a non-openshift
+	// namespace, will be under the key ""
+	nsIntervals := map[string][][]interface{}{}
 
 	for _, i := range intervals {
-		logLine := map[string]string{
-			"_entry": i.Message,
-		}
+		namespace, logLine, err := intervalToLogLine(i, timeSuffix)
 
-		if strings.HasPrefix(i.Locator, "e2e-test/\"") {
-			startIndex := strings.Index(i.Locator, "\"") + 1
-			endIndex := strings.LastIndex(i.Locator, "\"")
-			logLine["e2e-test"] = i.Locator[startIndex:endIndex]
+		// Convert the log data to a packed json log line, where the message is in '_entry'.
+		// This allows us to search additional fields we didn't index as labels using the unpack/json
+		// directives in logQL.
+		logLineJson, err := json.Marshal(logLine)
+		if err != nil {
+			logrus.WithError(err).Errorf("unable to mashall log line to json: %s", logLine)
 			continue
 		}
 
+		if _, ok := nsIntervals[namespace]; !ok {
+			nsIntervals[namespace] = make([][]interface{}, 0)
+		}
+		nsIntervals[namespace] = append(nsIntervals[namespace], []interface{}{
+			strconv.FormatInt(i.From.UnixNano(), 10),
+			string(logLineJson),
+		})
+
+		if len(nsIntervals[namespace]) == batchSize {
+			err := pushBatch(client, bearerToken, invoker, namespace, nsIntervals[namespace], dryRun)
+			if err != nil {
+				return err
+			}
+			nsIntervals[namespace] = make([][]interface{}, 0)
+		}
+	}
+
+	// Push the final batches for each namespace:
+	logrus.Info("pushing final batches for each remaining namespace")
+	for namespace, v := range nsIntervals {
+		if len(v) > 0 {
+			err := pushBatch(client, bearerToken, invoker, namespace, nsIntervals[namespace], dryRun)
+			if err != nil {
+				return err
+			}
+			nsIntervals[namespace] = make([][]interface{}, 0)
```

Comment:
since you've pushed the rest, do we need this line?

Reply:
but it is harmless

Reply:
Nope looks like you're right, probably didn't need that line but indeed harmless.